### PR TITLE
Fix generateXAddress() and generateXAddress() with no entropy

### LINKED ISSUE #1209: generateXAddress - Cannot read property 'Symbol(Symbol.iterator)' of undefined
rippleHelper.js
Screen Shot 2020-02-14 at 10 02 40 AM
test.js
Screen Shot 2020-02-14 at 10 03 49 AM

Run "node test.js"
> (node:65181) UnhandledPromiseRejectionWarning: UnexpectedError: Cannot read property 'Symbol(Symbol.iterator)' of undefined
    at Object.generateAddressAPI (/Users/vanphandinh/Desktop/projects/blockchain/node_modules/ripple-lib/dist/npm/offline/generate-address.js:29:15)
    at RippleAPI.generateXAddress (/Users/vanphandinh/Desktop/projects/blockchain/node_modules/ripple-lib/dist/npm/api.js:219:35)
    at RippleHelper.generateAddress (/Users/vanphandinh/Desktop/projects/blockchain/src/ripple2/rippleHelper.js:37:25)
    at test (/Users/vanphandinh/Desktop/projects/blockchain/src/ripple2/test.js:30:25)
    at Object.<anonymous> (/Users/vanphandinh/Desktop/projects/blockchain/src/ripple2/test.js:45:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
(node:65181) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:65181) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

### PR DESCRIPTION
**Pull request by intelliot**
Fix #1209

Calling:
- `Uint8Array.from(undefined)`

Throws:
- `TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at Function.from (<anonymous>)`

Test coverage was poor, so I added a ton of new tests.

### CODE REVIEW

File: test/api/generateAddress/index.ts [hunk 1]
```diff
@@ -1,30 +1,209 @@
 import assert from 'assert-diff'
 import responses from '../../fixtures/responses'
 import {TestSuite} from '../../utils'
+import { GenerateAddressOptions } from '../../../src/offline/generate-address'
 const {generateAddress: RESPONSE_FIXTURES} = responses
 
+export const config = {
+  // These tests do not use `address` so without skipping the address iteration, we would just
+  // redundantly run the same test twice, with no changes.
+  // Known issue: The test runner will say that these tests are being run under `[Original Address]`.
```
Comment (hbergren) [inline test/api/generateAddress/index.ts]: What causes this issue?

File: src/offline/generate-address.ts [hunk 2]
```diff
@@ -28,10 +28,10 @@ export interface GenerateAddressOptions {
...
   validate.generateAddress({options})
   try {
-    const secret = keypairs.generateSeed({
-      entropy: Uint8Array.from(options.entropy),
-      algorithm: options.algorithm
-    })
+    if (options && options.entropy) {
+      options.entropy = Uint8Array.from(options.entropy)
+    }
+    const secret = keypairs.generateSeed(options as { entropy?: Uint8Array; algorithm?: "ecdsa-secp256k1" | "ed25519"; })
```
[the remark this answers is not in the document]
Comment (intelliot) [reply src/offline/generate-address.ts]:
```
ripple-lib/node_modules/ts-node/src/index.ts:421
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
src/offline/generate-address.ts:34:42 - error TS2345: Argument of type 'GenerateAddressOptions' is not assignable to parameter of type '{ entropy?: Uint8Array; algorithm?: "ecdsa-secp256k1" | "ed25519"; }'.
  Types of property 'entropy' are incompatible.
    Type 'Uint8Array | number[]' is not assignable to type 'Uint8Array'.
      Type 'number[]' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteLength, byteOffset, and 3 more.

34     const secret = keypairs.generateSeed(options)
                                            ~~~~~~~

    at createTSError (/Users/elliotlee/ripple/ripple-lib/node_modules/ts-node/src/index.ts:421:12)
    at reportTSError (/Users/elliotlee/ripple/ripple-lib/node_modules/ts-node/src/index.ts:425:19)
    at getOutput (/Users/elliotlee/ripple/ripple-lib/node_modules/ts-node/src/index.ts:530:36)
    at Object.compile (/Users/elliotlee/ripple/ripple-lib/node_modules/ts-node/src/index.ts:735:32)
[...]
```

File: test/api/generateAddress/index.ts (hunk 1 above)
Comment (intelliot) [reply test/api/generateAddress/index.ts]: https://github.com/ripple/ripple-lib/blob/develop/test/ripple-api-test.ts#L36

We currently assume that every method takes an address. I have an idea for fixing this; I might put it in a separate PR.

File: src/offline/generate-address.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (keefertaylor) [reply src/offline/generate-address.ts]: It's generally a bad idea to mutate your input parameters (it makes your code hard to reason about). In this instance, it also forces you to have an ugly cast :) 

Just make a new `shimOptions` object and assign the uint8array on that. 



https://spin.atomicobject.com/2011/04/10/javascript-don-t-reassign-your-function-arguments/

File: test/api/generateAddress/index.ts (hunk 1 above)
Comment (intelliot) [reply test/api/generateAddress/index.ts]: Nevermind - I fixed it here 🎉 

It turns out there were a bunch of other tests that also were being run repeatedly, unnecessarily. Now tests are run once, unless they take an address, in which case they are run first with a classic address and second with an X-address, as intended.

File: test/api/generateAddress/index.ts [hunk 3]
```diff
@@ -9,22 +10,192 @@ const {generateAddress: RESPONSE_FIXTURES} = responses
...
+  },
+
+  'generateAddress with algorithm `ed25519`': async (api) => {
+    // GIVEN we want to use 'ed25519'
+    const options: GenerateAddressOptions = {algorithm: 'ed25519'}
+
+    // WHEN generating an address
+    const account = api.generateAddress(options)
+
+    // THEN we get an object with an address starting with 'r' and a secret starting with 's'
```
Comment (hbergren) [inline test/api/generateAddress/index.ts]: `THEN we get... a secret starting with 'sEd'`?

(For all the `ed25519` tests?)

**Pull request merged**